src/par_iter/find_first_last.rs

```diff
@@ -0,0 +1,183 @@
+    }
+
+    fn into_folder(self) -> Self::Folder {
+        FindFolder {
+            find_op: self.find_op,
+            boundary: match self.match_position {
+                MatchPosition::Leftmost => self.lower_bound.get(),
+                MatchPosition::Rightmost => self.upper_bound
+            },
+            match_position: self.match_position,
+            best_found: self.best_found,
+            item: None,
+        }
+    }
+
+    fn full(&self) -> bool {
+        let best = self.best_found.load(Ordering::Relaxed);
+        match self.match_position {
+            // can stop consuming if the best found index so far is *strictly*
+            // better than anything this consumer will find
+            MatchPosition::Leftmost => best < self.lower_bound.get(),
+            MatchPosition::Rightmost => best > self.upper_bound
+        }
+    }
+}
+
+impl<'f, ITEM, FIND_OP> UnindexedConsumer<ITEM> for FindConsumer<'f, FIND_OP>
+    where ITEM: Send,
+          FIND_OP: Fn(&ITEM) -> bool + Sync
+{
+    fn split_off(&self) -> Self {
+        // Upper bound for one consumer will be lower bound for the other. This
+        // overlap is okay, because only one of the bounds will be used for
+        // comparing against best_found; the other is kept only to be able to
+        // divide the range in half
+        let old_lower_bound = self.lower_bound.get();
+        let median = old_lower_bound + ((self.upper_bound - old_lower_bound) / 2);
+        self.lower_bound.set(median);
+
+        FindConsumer {
+            find_op: self.find_op,
+            lower_bound: Cell::new(old_lower_bound),
+            upper_bound: median,
+            match_position: self.match_position,
+            best_found: self.best_found,
+        }
+    }
+
+    fn to_reducer(&self) -> Self::Reducer {
+        FindReducer { match_position: self.match_position }
+    }
+}
+
+struct FindFolder<'f, ITEM, FIND_OP: 'f> {
+    find_op: &'f FIND_OP,
+    boundary: usize,
+    match_position: MatchPosition,
+    best_found: &'f AtomicUsize,
+    item: Option<ITEM>,
+}
+
+impl<'f, FIND_OP: 'f + Fn(&ITEM) -> bool, ITEM> Folder<ITEM> for FindFolder<'f, ITEM, FIND_OP> {
+    type Result = Option<ITEM>;
+
+    fn consume(mut self, item: ITEM) -> Self {
+        if (self.find_op)(&item) {
+            // This may sometimes set best_found to a worse index than it was
+            // before, depending on timing. This means more consumers will
+            // continue to run than necessary, but the reducer will still ensure
+            // the correct value is returned.
+            self.best_found.swap(self.boundary, Ordering::Relaxed);
+            self.item = Some(item);
+        }
+        self
+    }
+
+    fn complete(self) -> Self::Result {
+        self.item
+    }
+
+    fn full(&self) -> bool {
+        let best_found = self.best_found.load(Ordering::Relaxed);
+        match self.match_position {
+            MatchPosition::Leftmost => best_found < self.boundary,
```

Comment:
@nikomatsakis I think he's saying it's *also* sufficient, which is correct (and a case I didn't think about). If we've already found the item, we need to make sure we stop before we find some other item slightly to the right.

So @cuviper, I think you're right that that logic should be added to both `full` and `consume`.